Move best practices out of "Type Stubs" doc into guide on writing stubs. (#1817)
python/typing: docs/guides/writing_stubs.rst

If your package has some particularly complex aspects, you could even consider
writing dedicated typing tests for tricky definitions. For more details, see
:ref:`testing`.

Stub Content
============

This section documents best practices on what elements to include or
leave out of stub files.

Modules excluded fom stubs
--------------------------

> Not all modules should be included into stubs.

Reviewer: Not a native speaker, but "into stubs" sounds wrong to be. "in stubs"?

Author: Fixed, thanks!

clear exactly what is part of the interface.

The following should always be included:

* All objects listed in the module's documentation.
* All objects included in ``__all__`` (if present).

Other objects may be included if they are not prefixed with an underscore
or if they are being used in practice. (See the next section.)

.. _undocumented-objects:

Undocumented Objects
--------------------

Undocumented objects may be included as long as they are marked with a comment
of the form ``# undocumented``.

Example::

    def list2cmdline(seq: Sequence[str]) -> str: ...  # undocumented

Such undocumented objects are allowed because omitting objects can confuse
users. Users who see an error like "module X has no attribute Y" will
not know whether the error appeared because their code had a bug or
because the stub is wrong. Although it may also be helpful for a type
checker to point out usage of private objects, false negatives (no errors for
wrong code) are preferable over false positives (type errors
for correct code). In addition, even for private objects a type checker
can be helpful in pointing out that an incorrect type was used.

``__all__``
------------

> A stub file should contain an ``__all__`` variable if and only if it also

Reviewer 2: ```suggestion
A stub file should contain an ``__all__`` variable if and only if it is also
```

present at runtime. In that case, the contents of ``__all__`` should be
identical in the stub and at runtime. If the runtime dynamically adds
or removes elements (for example if certain functions are only available on

> some platforms), include all possible elements in the stubs.

Reviewer 2: ... or use sys.platform checks to replicate the runtime logic. Let's use an example that can't be represented in stubs:
```suggestion
some system configurations), include all possible elements in the stubs.
```

wrong code) are preferable over false positives (type errors
for correct code). In addition, even for private objects a type checker
can be helpful in pointing out that an incorrect type was used.

``__all__``
------------

A stub file should contain an ``__all__`` variable if and only if it also
present at runtime. In that case, the contents of ``__all__`` should be
identical in the stub and at runtime. If the runtime dynamically adds
or removes elements (for example if certain functions are only available on
some platforms), include all possible elements in the stubs.

Stub-Only Objects
-----------------

Definitions that do not exist at runtime may be included in stubs to aid in
expressing types. Sometimes, it is desirable to make a stub-only class available
to a stub's users — for example, to allow them to type the return value of a
public method for which a library does not provided a usable runtime type. Use
the ``typing.type_check_only`` decorator to mark such objects::

  from typing import Protocol, type_check_only

  @type_check_only
  class Readable(Protocol):
      def read(self) -> str: ...

  def get_reader() -> Readable: ...

Structural Types
----------------

> As seen in the example with ``_Readable`` in the previous section, a common use

Reviewer 2: ```suggestion
As seen in the example with ``Readable`` in the previous section, a common use
```

of stub-only objects is to model types that are best described by their
structure. These objects are called protocols (:pep:`544`), and it is encouraged
to use them freely to describe simple structural types.

> It is `recommended <#private-definitions>`_ to prefix stub-only object names with ``_``.

Author: I deleted this sentence and reworded the "Stub-Only Objects" section to say that names should be underscore-prefixed unless you want to expose them to users, in which case use `@type_check_only`.

Incomplete Stubs
----------------

Partial stubs can be useful, especially for larger packages, but they should
follow the following guidelines:

* Included functions and methods should list all arguments, but the arguments
  can be left unannotated.
* Do not use ``Any`` to mark unannotated or partially annotated values. Leave
  function parameters and return values unannotated and mark the function with

>   an ``# incomplete`` comment. This comment is mainly intended as a reminder for

Author: Delorted.

      def __setattr__(self, name: str, value: int) -> None: ...

``__delattr__`` should not be included in stubs.

Finally, even in the presence of ``__getattr__`` and ``__setattr__``, it is
still recommended to separately define known attributes.

Constants
---------

When the value of a constant is important,  mark it as ``Final`` and assign it
to its value.

Yes::

    TEL_LANDLINE: Final = "landline"
    TEL_MOBILE: Final = "mobile"
    DAY_FLAG: Final = 0x01
    NIGHT_FLAG: Final = 0x02

No::

    TEL_LANDLINE: str
    TEL_MOBILE: str
    DAY_FLAG: int
    NIGHT_FLAG: int

Documentation or Implementation
-------------------------------

Sometimes a library's documented types will differ from the actual types in the
code. In such cases, stub authors should use their best judgment. Consider these
two examples::

  def print_elements(x):
      """Print every element of list x."""
      for y in x:
          print(y)

  def maybe_raise(x):
      """Raise an error if x (a boolean) is true."""
      if x:
          raise ValueError()

The implementation of ``print_elements`` takes any iterable, despite the
documented type of ``list``. In this case, annotate the argument as

> ``Iterable[Any]``, to follow the :ref:`best practice<argument-return-practices>`

Reviewer 2: ```suggestion
``Iterable[object]``, to follow the :ref:`best practice<argument-return-practices>`
```

bools, ints, bytes, strings, and floats). Use the ellipsis literal ``...`` in
place of more complex default values. Use an explicit ``X | None`` annotation
when the default is ``None``.

Yes::

    def foo(x: int = 0) -> None: ...
    def bar(y: str | None = None) -> None: ...

No::

    def foo(x: X = X()) -> None: ...
    def bar(y: str = None) -> None: ...

Do not annotate ``self`` and ``cls`` in method definitions, except when
referencing a type variable.

Yes::

    _T = TypeVar("_T")

    class Foo:
        def bar(self) -> None: ...
        @classmethod
        def create(cls: type[_T]) -> _T: ...

No::

    class Foo:
        def bar(self: Foo) -> None: ...
        @classmethod
        def baz(cls: type[Foo]) -> int: ...

The bodies of functions and methods should consist of only the ellipsis
literal ``...`` on the same line as the closing parenthesis and colon.

Yes::

    def to_int1(x: str) -> int: ...
    def to_int2(
        x: str,
    ) -> int: ...

No::

    def to_int1(x: str) -> int:
        return int(x)
    def to_int2(x: str) -> int:
        ...
    def to_int3(x: str) -> int: pass

.. _private-definitions:

> Private Definitions

Reviewer 2: Duplicates a previous section

Author: I consolidated this with the "Stub-Only Objects" section.
